Add stm32u3 series

kbidani opened the pull request:
Add stm32u3 series

kbidani committed:
soc: st: stm32: Add series stm32u3

Add STM32U3 familly support

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

kbidani committed:
dts: arm: st: add stm32u385 dtsi files

Provide support for the ST32U385 series

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

nordicjm commented on soc/st/stm32/soc.yml:
```diff
@@ -199,6 +199,9 @@ family:
     - name: stm32u031xx
```
commit typo: *series

kbidani committed:
boards: st: add nucleo_u385rg_q board support

add board support for nucleo_u385rg_q

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

kbidani committed:
include: zephyr: Add the stm32u3 clock control and the reset register

Add the stm32u3 clock control and the RCC bus reset register offset

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

kbidani committed:
drivers: clock_control: Add STM32U3XX clock support

add clock support for STM32U3X SoC series.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

kbidani committed:
dts: bindings: clock: add STM32U3 MSI

Add the Device Tree bindings for the
MSI clock of STM32U3 series.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

erwango commented on soc/st/stm32/stm32u3x/Kconfig:
```diff
@@ -0,0 +1,15 @@
+	select CPU_CORTEX_M33
+	select CPU_HAS_ARM_SAU
+	select CPU_HAS_ARM_MPU
+	select CPU_HAS_FPU
+	select ARMV8_M_DSP
```
Please confirm that FPU and DSP are really compatible with U3

kbidani replied:
Yes, the compatibility of the FPU and DSP with the U3 series is confirmed in the include/stm32u385xx.h file, specifically in the section detailing the configuration of the Cortex-M33 Processor and Core Peripherals:
#define __FPU_PRESENT                   1U        /* FPU present */
#define __DSP_PRESENT                   1U        /* DSP extension present */

erwango commented on soc/st/stm32/stm32u3x/Kconfig:
```diff
@@ -0,0 +1,15 @@
+# ST Microelectronics STM32U3 MCU series
+
+# Copyright (c) 2025 Linaro Limited
```
```suggestion
# Copyright (c) 2025 STMicroelectronics
```
Please fix everywhere

erwango commented on west.yml:
```diff
@@ -243,7 +243,7 @@ manifest:
         - hal
     - name: hal_stm32
-      revision: 6e4716f8ac2c6e6159a958a836e1dcca06804456
+      revision: 272
```
```
      revision: pull/272/head
```

erwango commented on drivers/clock_control/CMakeLists.txt:
```diff
@@ -75,8 +75,8 @@ elseif(CONFIG_SOC_SERIES_STM32H5X)
   zephyr_library_sources(clock_stm32_ll_n6.c)
-elseif(CONFIG_SOC_SERIES_STM32U5X)
-  zephyr_library_sources(clock_stm32_ll_u5.c)
+elseif(CONFIG_SOC_SERIES_STM32U5X OR CONFIG_SOC_SERIES_STM32U3X)
+  zephyr_library_sources(clock_stm32_ll_u5_u3.c)
```
```suggestion
elseif(CONFIG_SOC_SERIES_STM32U5X OR CONFIG_SOC_SERIES_STM32U3X)
  zephyr_library_sources(clock_stm32_ll_u5.c)
```

erwango commented on drivers/clock_control/clock_stm32_ll_u5_u3.c:
Keep this file name (clock_stm32_ll_u5.c).
One of the issue is that it shows as a new file which makes it more complex to review.
Also updating name is not required.

kbidani replied:
A new file clock_stm32_ll_u3.c has been created to avoid using #ifdef.

nordicjm commented on dts/arm/st/u3/stm32u3.dtsi:
```diff
@@ -0,0 +1,244 @@
+			clock-frequency = <DT_FREQ_K(32)>;
+			status = "disabled";
+		};
+
```
```suggestion
```

nordicjm replied:
not fixed

kbidani replied:
issue has been fixed now

nordicjm commented on dts/arm/st/u3/stm32u3.dtsi:
```diff
@@ -0,0 +1,244 @@
+				reg = <0x42021c00 0x400>;
+				clocks = <&rcc STM32_CLOCK(AHB2, 7U)>;
+			};
+
```
```suggestion
```

nordicjm replied:
also not fixed

kbidani replied:
issue has been fixed now

kbidani committed:
drivers: gpio: fix gpio warning

Implement conditional compilation to avoid the warning:
- Use LL_PWR_EnableVDDIO2() for STM32U3 series.
- Use LL_PWR_EnableVddIO2() for other series.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

nordicjm commented on boards/st/nucleo_u385rg/doc/img/nucleo_u385rg.webp:
does the resolution of this have to be so big? Non-blocking comment but would be nice to reduce

etienne-lms commented on boards/st/nucleo_u385rg/nucleo_u385rg_defconfig:
```diff
@@ -0,0 +1,17 @@
+CONFIG_GPIO=y
+
+# Enable Clocks
+CONFIG_CLOCK_CONTROL=y
```
Are `CONFIG_GPIO=y` and `CONFIG_CLOCK_CONTROL=y` really needed? Ain't they implicit?

erwango replied:
`CONFIG_CLOCK_CONTROL=y` could be removed, but `CONFIG_GPIO=y` is indeed required.

kbidani replied:
CONFIG_CLOCK_CONTROL to be removed but CONFIG_GPIO is required.

etienne-lms reviewed (commented):
Some nitpicking comments.

kbidani commented on boards/st/nucleo_u385rg/doc/index.rst:
```diff
@@ -0,0 +1,293 @@
+  capability: mass storage, Virtual COM port, and debug port
+- Flexible board power supply:
+
+  - USB VBUS or external source(3.3V, 5V, 7 - 12V)
```
The indentation is here to align with the colon and the line break.

etienne-lms replied:
Ok sorry, my mistake.

kbidani commented:
> Some nitpicking comments.

thanks^^

kbidani commented on boards/st/nucleo_u385rg/doc/index.rst:
fixed everywhere.

mathieuchopstm commented on boards/st/nucleo_u385rg/board.cmake:
```diff
@@ -0,0 +1,11 @@
+# keep first
+board_runner_args(stm32cubeprogrammer "--port=swd" "--reset-mode=hw")
+
+board_runner_args(pyocd "--target=stm32u385rgtxq")
+
+board_runner_args(jlink "--device=STM32U385RG" "--reset-after-load")
```
Unnecessary newlines?
```suggestion
board_runner_args(stm32cubeprogrammer "--port=swd" "--reset-mode=hw")
board_runner_args(pyocd "--target=stm32u385rgtxq")
board_runner_args(jlink "--device=STM32U385RG" "--reset-after-load")
```

mathieuchopstm commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+
+
+/* Macros to fill up prescaler values */
+#define z_ahb_prescaler(v) LL_RCC_HCLK_SYSCLK_DIV_ ## v
+#define ahb_prescaler(v) z_ahb_prescaler(v)
+
+#define z_apb1_prescaler(v) LL_RCC_APB1_HCLK_DIV_ ## v
+#define apb1_prescaler(v) z_apb1_prescaler(v)
+
+#define z_apb2_prescaler(v) LL_RCC_APB2_HCLK_DIV_ ## v
+#define apb2_prescaler(v) z_apb2_prescaler(v)
+
+#define z_apb3_prescaler(v) LL_RCC_APB3_HCLK_DIV_ ## v
+#define apb3_prescaler(v) z_apb3_prescaler(v)
```
Is the double macro layer really necessary?

mathieuchopstm replied:
What about using Zephyr's `CONCAT` from `<zephyr/util/sys.h>`? Maybe the following would work?
```
#define ahb_prescaler(v) CONCAT(LL_RCC_HCLK_SYSCLK_DIV_, v)
```

mathieuchopstm commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+#define z_apb3_prescaler(v) LL_RCC_APB3_HCLK_DIV_ ## v
+#define apb3_prescaler(v) z_apb3_prescaler(v)
+
+static uint32_t get_bus_clock(uint32_t clock, uint32_t prescaler)
+{
+	return clock / prescaler;
+}
```
Considering how trivial this is, it would be easier to replace calls to this with the division directly (IMO).

mathieuchopstm commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+	{
+	frequency = MSIRC0_VALUE;
+	}
+	switch (LL_RCC_MSIS_GetClockDivision())
+	{
+	case LL_RCC_MSIS_CLOCK_SOURCE_RC_DIV_2:
+	frequency /= 2U;
+	break;
+	case LL_RCC_MSIS_CLOCK_SOURCE_RC_DIV_4:
+	frequency /= 4U;
+	break;
+	case LL_RCC_MSIS_CLOCK_SOURCE_RC_DIV_8:
+	frequency /= 8U;
+	break;
+	case LL_RCC_MSIS_CLOCK_SOURCE_RC_DIV_1:
+	break;
+	default:
+	return 0;
+	}
```
Indentation
```suggestion
	switch (LL_RCC_MSIS_GetClockDivision())
	{
	case LL_RCC_MSIS_CLOCK_SOURCE_RC_DIV_2:
		frequency /= 2U;
		break;
	case LL_RCC_MSIS_CLOCK_SOURCE_RC_DIV_4:
		frequency /= 4U;
		break;
	case LL_RCC_MSIS_CLOCK_SOURCE_RC_DIV_8:
		frequency /= 8U;
		break;
	case LL_RCC_MSIS_CLOCK_SOURCE_RC_DIV_1:
		break;
	default:
		return 0;
	}
```

etienne-lms replied:
nitpicking: `switch (LL_RCC_MSIS_GetClockDivision()) {` at first switch/case line.

Also: I suggest to  implement `return frequency / <value>` straight in the cases, instead of `frequency /= <value>; break;`.

mathieuchopstm commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+{
+	uint32_t frequency = MSIRC1_VALUE;
+
+	if (LL_RCC_MSIS_GetClockSource() == LL_RCC_MSIS_CLOCK_SOURCE_RC0)
+	{
+	frequency = MSIRC0_VALUE;
+	}
```
Indentation + braces
```suggestion
	if (LL_RCC_MSIS_GetClockSource() == LL_RCC_MSIS_CLOCK_SOURCE_RC0) {
		frequency = MSIRC0_VALUE;
	}
```

mathieuchopstm commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+	}
+
+	if (IS_ENABLED(STM32_LSE_ENABLED)) {
+		/* Enable the power interface clock */
+
+		LL_AHB1_GRP2_EnableClock(LL_AHB1_GRP2_PERIPH_PWR);
```
Already done in SoC init hook.
```suggestion
		/* N.B.: PWR clock has been enabled by SoC init hook */
```

mathieuchopstm commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+		}
+
+		/* Enable Epod Booster */
+		LL_AHB1_GRP2_EnableClock(LL_AHB1_GRP2_PERIPH_PWR);
```
Already done in SoC init hook.
```suggestion
```

mathieuchopstm commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+	}
+
+	if (IS_ENABLED(STM32_LSI_ENABLED)) {
+		if (!LL_AHB1_GRP2_IsEnabledClock(LL_AHB1_GRP2_PERIPH_PWR)) {
+		/* Enable the power interface clock */
+		LL_AHB1_GRP2_EnableClock(LL_AHB1_GRP2_PERIPH_PWR);
+	}
```
Already done in SoC init hook.
```suggestion
```

mathieuchopstm commented on dts/arm/st/u3/stm32u3.dtsi:
```diff
@@ -0,0 +1,241 @@
+			compatible = "st,stm32u5-rcc";
+			clocks-controller;
+			#clock-cells = <2>;
+			reg = < 0x40030c00 0x400>;
```
```suggestion
			reg = <0x40030c00 0x400>;
```

mathieuchopstm replied:
.

etienne-lms replied:
To address.

mathieuchopstm commented on dts/arm/st/u3/stm32u3.dtsi:
```diff
@@ -0,0 +1,241 @@
+				gpio-controller;
+				#gpio-cells = <2>;
+				reg = <0x42020000 0x400>;
+				clocks = <&rcc STM32_CLOCK(AHB2, 0U)>;
```
Do not use `U` suffix in DTS. Ditto for all nodes.
```suggestion
				clocks = <&rcc STM32_CLOCK(AHB2, 0)>;
```

mathieuchopstm commented on dts/arm/st/u3/stm32u385Xg.dtsi:
```diff
@@ -0,0 +1,30 @@
+#include <st/u3/stm32u385.dtsi>
+
+/ {
+	sram1: memory@20000000 {
+		compatible = "zephyr,memory-region", "mmio-sram";
+		reg = <0x20000000 DT_SIZE_K(192)>;
+		zephyr,memory-region = "SRAM1";
+	};
+
+	sram2: memory@20030000 {
+		compatible = "zephyr,memory-region", "mmio-sram";
+		reg = <0x20030000 DT_SIZE_K(64)>;
+		zephyr,memory-region = "SRAM2";
+	};
```
I think contiguous regions are usually merged into a single one on other series?

mathieuchopstm commented on dts/bindings/clock/st,stm32u3-msi-clock.yaml:
```diff
@@ -0,0 +1,29 @@
+    default: 0
+    required: true
+    type: int
+    description: |
+      MSI clock ranges
```
Nit: no need for multi-line description if it's small
```suggestion
    description: MSI clock ranges
```

mathieuchopstm commented on include/zephyr/dt-bindings/reset/stm32u3_reset.h:
```diff
@@ -0,0 +1,19 @@
+#include "stm32-common.h"
+
+/* RCC bus reset register offset */
+#define STM32_RESET_BUS_AHB1  0x060
+#define STM32_RESET_BUS_APB1L 0x074
+#define STM32_RESET_BUS_APB1H 0x078
+#define STM32_RESET_BUS_APB2  0x07C
+#define STM32_RESET_BUS_APB3  0x080
```
Could you add the missing AHB2L/AHB2H?

mathieuchopstm commented on soc/st/stm32/stm32u3x/Kconfig.defconfig.stm32u385xx:
Unless it's against some Zephyr guideline: as far as I can tell, all STM32U3 variants have 125 interrupts, so this could be integrated in the series-wide `Kconfig.defconfig`. IMO this is better as it avoids having 150 copies of the same file, except with a different SKU in the `if` clause, and makes introducing other lines simpler (c.f. STM32WB0 series).
cc @erwango what do you think?

mathieuchopstm reviewed (changes requested):
Some "must change" and some remarks/questions

mathieuchopstm commented:
\+ general remark about `SYS_CLOCK_HW_CYCLES_PER_SEC`: as documented in Kconfig, this should `[specify] the frequency of the hardware timer used for the system clock (in Hz)`, where *system clock* is to be interpreted as *the driver/device implementing the `sys_clock_XXX` API*.

On most STM32, we use SysTick as System Clock and it runs at the same frequency as the CPU, so `SYS_CLOCK_HW_CYCLES_PER_SEC` can be used as an ad-hoc way of reading the CPU frequency, but this is an implementation detail that may not hold true (c.f. [STM32MP13](https://github.com/zephyrproject-rtos/zephyr/blob/74056a7aedd466b455ec1440c34a79f4e659ddff/soc/st/stm32/stm32mp13x/Kconfig.defconfig#L13-L18)). This could be an issue with low-power modes since we replace the System Clock driver. Maybe another mechanism should be used, although this discussion is probably outside PR scope.

etienne-lms commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -417,7 +419,7 @@
 #endif
 
-#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msis), st_stm32u5_msi_clock, okay)
+#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msis), st_stm32u5_msi_clock, okay) || DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msis), st_stm32u3_msi_clock, okay)
```
```suggestion
#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msis), st_stm32u5_msi_clock, okay) || \
	DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msis), st_stm32u3_msi_clock, okay)

etienne-lms commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -427,7 +429,7 @@
 #endif
 
-#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msik), st_stm32u5_msi_clock, okay)
+#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msik), st_stm32u5_msi_clock, okay) || DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msik), st_stm32u3_msi_clock, okay)
```
ditto
```suggestion
#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msik), st_stm32u5_msi_clock, okay) || \
	DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msik), st_stm32u3_msi_clock, okay)

etienne-lms commented on include/zephyr/dt-bindings/clock/stm32u3_clock.h:
```diff
@@ -0,0 +1,103 @@
+
+/** Domain clocks */
+
+/* RM0456, Figure 36 Clock tree for STM32U3 Series */
```
RM0487

etienne-lms commented on include/zephyr/dt-bindings/clock/stm32u3_clock.h:
```diff
@@ -0,0 +1,103 @@
+#define STM32_PERIPH_BUS_MIN	STM32_CLOCK_BUS_AHB1
+#define STM32_PERIPH_BUS_MAX	STM32_CLOCK_BUS_APB3
+
+/** @brief RCC_CCIPRx register offset (RM0456.pdf) */
```
RM0487

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+	    ((src_clk == STM32_SRC_LSI) && IS_ENABLED(STM32_LSI_ENABLED)) ||
+	    ((src_clk == STM32_SRC_MSIS) && IS_ENABLED(STM32_MSIS_ENABLED)) ||
+	    ((src_clk == STM32_SRC_MSIK) && IS_ENABLED(STM32_MSIK_ENABLED))
+	) {
```
```patch
-	    ((src_clk == STM32_SRC_MSIK) && IS_ENABLED(STM32_MSIK_ENABLED))
-	) {
+	    ((src_clk == STM32_SRC_MSIK) && IS_ENABLED(STM32_MSIK_ENABLED))) {

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+}
+
+/** @brief Verifies clock is part of active clock configuration */
+int enabled_clock(uint32_t src_clk)
```
`static`

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+	}
+
+	sys_set_bits(DT_REG_ADDR(DT_NODELABEL(rcc)) + pclken->bus,
+		     pclken->enr);
```
nitpicking: fits in a single line:
```c
	sys_set_bits(DT_REG_ADDR(DT_NODELABEL(rcc)) + pclken->bus, pclken->enr);

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+	}
+
+	sys_clear_bits(DT_REG_ADDR(DT_NODELABEL(rcc)) + pclken->bus,
+		       pclken->enr);
```
ditto:
```c
	sys_clear_bits(DT_REG_ADDR(DT_NODELABEL(rcc)) + pclken->bus, pclken->enr);

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+
+static int stm32_clock_control_on(const struct device *dev, clock_control_subsys_t sub_system)
+{
+	struct stm32_pclken *pclken = (struct stm32_pclken *)(sub_system);
```
Could remove the parenthesis around `sub_system`?
Ditto in the other local functions below.

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+	}
+
+	if (pclken->div) {
+		*rate /= (pclken->div + 1);
```
Can remove parentheses.

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+
+static void set_up_fixed_clock_sources(void)
+{
+
```
Could you remove this empty line?

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+		/* Enable HSE */
+		LL_RCC_HSE_Enable();
+		while (LL_RCC_HSE_IsReady() != 1) {
+		/* Wait for HSE ready */
```
indentation

Ditto at line 340.

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+		else if(STM32_MSIS_RANGE >= 4 && STM32_MSIS_RANGE <= 7)
+		{
+			LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE2);
+		}
```
```c
		while (LL_FLASH_GetLatency() != LL_FLASH_LATENCY_2) {
		}
		if (STM32_MSIS_RANGE >= 0U && STM32_MSIS_RANGE <= 3U) {
			LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE1);
		} else if(STM32_MSIS_RANGE >= 4U && STM32_MSIS_RANGE <= 7U) {
			LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE2);
		}

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+		LL_PWR_EnableEPODBooster();
+		while(LL_PWR_IsActiveFlag_BOOST() != 1)
+		{
+		}
```
```patch
-		while(LL_PWR_IsActiveFlag_BOOST() != 1)
+		while (LL_PWR_IsActiveFlag_BOOST() != 1U) {
-		{
```

Nitpicking: I would suggest: `while (LL_PWR_IsActiveFlag_BOOST() == 0U)`, for all while loop expecting to test a 0 or non-zero value.

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+		/* Activate MSIS as source */
+		__HAL_RCC_MSIS_ENABLE();
+		if(STM32_MSIS_RANGE >= 0 && STM32_MSIS_RANGE <= 3)
+		{
```
also missing a space char after `if`, move `{` to the end of the previous line.

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+		{
+			LL_RCC_MSIS_SetClockSource(LL_RCC_MSIS_CLOCK_SOURCE_RC0);
+			switch (STM32_MSIS_RANGE)
+			{
```
Ditto here + indent the cases instruction sequences below.

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+		/* Set HSE as SYSCLCK source */
+		LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_HSE);
+		while (LL_RCC_GetSysClkSource() != LL_RCC_SYS_CLKSOURCE_STATUS_HSE) {
+		}
```
Remove 1 indentation tab, above and below.

etienne-lms commented on dts/bindings/clock/st,stm32u3-msi-clock.yaml:
```diff
@@ -0,0 +1,29 @@
+
+  msi-range:
+    default: 0
+    required: true
```
Can remove `required` if there's a `default` value.

etienne-lms commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -723,7 +723,11 @@ static int gpio_stm32_init(const struct device *dev)
 	z_stm32_hsem_lock(CFG_HW_RCC_SEMID, HSEM_LOCK_DEFAULT_RETRY);
 	/* Port G[15:2] requires external power supply */
 	/* Cf: L4/L5 RM, Chapter "Independent I/O supply rail" */
```
```suggestion
	/* Cf: L4/L5/U3 RM, Chapter "Independent I/O supply rail" */

FRASTM commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,603 @@
+			LL_RCC_MSI_RC1_PLLmode_Enable();
+			while (!LL_RCC_MSI_RC1_PLLmode_IsEnabled()) {
+			}
+		}
```
It seems that HSE and HSE/2 are also possible LL_RCC_MSI_RC0_SetPLLInputClk

kbidani commented on soc/st/stm32/stm32u3x/Kconfig.defconfig:
```diff
@@ -0,0 +1,10 @@
+# ST Microelectronics STM32U3 MCU line
```
+# STMicroelectronics STM32U3 MCU series

kbidani commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,587 @@
+
+		LL_AHB1_GRP2_EnableClock(LL_AHB1_GRP2_PERIPH_PWR);
+
+		if (!LL_PWR_IsEnabledBkUpAccess()) {
```
fixed everywhere.

erwango commented on west.yml:
```diff
@@ -243,7 +243,7 @@ manifest:
         - hal
     - name: hal_stm32
-      revision: 6e4716f8ac2c6e6159a958a836e1dcca06804456
+      revision: pull/272/head
```
Merged
```suggestion
      revision: a5537345c0e8f24841edfee00a68c28f339cea21
```

mathieuchopstm commented on boards/st/nucleo_u385rg_q/nucleo_u385rg_q_defconfig:
```diff
@@ -0,0 +1,14 @@
+# Enable MPU
+CONFIG_ARM_MPU=y
+
+# Serial Drivers
+CONFIG_SERIAL=y
```
For consistency
```suggestion
# Enable UART driver
CONFIG_SERIAL=y
```

mathieuchopstm commented on boards/st/nucleo_u385rg_q/nucleo_u385rg_q_defconfig:
```diff
@@ -0,0 +1,14 @@
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+
+# GPIO Controller
+CONFIG_GPIO=y
```
For consistency
```suggestion
# Enable GPIO driver
CONFIG_GPIO=y
```

mathieuchopstm commented on dts/arm/st/u3/stm32u385Xg.dtsi:
```diff
@@ -0,0 +1,26 @@
+#include <st/u3/stm32u385.dtsi>
+
+/ {
+	sram1: memory@20000000 {
```
It seems other series reuse the `sram0` nodelabel, rather than adding a "properly named" nodelabel. Unless `sram1` is used somewhere, I think we should
```suggestion
	sram0: memory@20000000 {
```

mathieuchopstm commented on soc/st/stm32/stm32u3x/soc.c:
```diff
@@ -0,0 +1,39 @@
+	/* Update CMSIS SystemCoreClock variable (HCLK) */
+	/* The MSIS is used as system clock source after startup from reset, configured at 12 MHz. */
+	SystemCoreClock = 12000000;
+
+	LL_AHB1_GRP2_EnableClock(LL_AHB1_GRP2_PERIPH_PWR);
```
Nit: documentation comment
```suggestion
        /* Enable power controller bus clock for other drivers */
	LL_AHB1_GRP2_EnableClock(LL_AHB1_GRP2_PERIPH_PWR);
```

mathieuchopstm commented on soc/st/stm32/stm32u3x/soc.c:
```diff
@@ -0,0 +1,39 @@
+
+	/* Update CMSIS SystemCoreClock variable (HCLK) */
+	/* The MSIS is used as system clock source after startup from reset, configured at 12 MHz. */
+	SystemCoreClock = 12000000;
+	/* Enable power controller bus clock for other drivers */
```
Nit: newline
```suggestion
	SystemCoreClock = 12000000;

	/* Enable power controller bus clock for other drivers */
```

mathieuchopstm reviewed (commented):
Besides this, [this other point I raised earlier](https://github.com/zephyrproject-rtos/zephyr/pull/88794#discussion_r2106930562) and other's comments, LGTM.

kbidani committed:
drivers: flash: add stm32u3 devices

Introduce the stm32u3 serie to the the existing flash driver
It is based on the stm32l5 model.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

gautierg-st commented on dts/bindings/clock/st,stm32u3-msi-clock.yaml:
```diff
@@ -0,0 +1,27 @@
+    type: int
+    description: MSI clock ranges
+    enum:
+      - 0 # range 0 around 96 MHz
+      - 1 # range 1 around 48 MHz
+      - 2 # range 2 around 24 MHz
+      - 3 # range 3 around 12 MHz
+      - 4 # range 4 around 24 MHz
+      - 5 # range 5 around 12 MHz
+      - 6 # range 6 around 6 MHz
+      - 7 # range 7 around 3 MHz
```
The comments are not visible in documentation. Please move them in the description.

kbidani committed:
dts: bindings: flash: add STM32U3 flash controller

Add the Device Tree binding for the STM32U3 flash controller.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

etienne-lms commented on dts/arm/st/u3/stm32u3.dtsi:
```diff
@@ -0,0 +1,241 @@
+		};
+
+		uart4: serial@40004c00 {
+			compatible = "st,stm32-usart", "st,stm32-uart";
```
```suggestion
			compatible = "st,stm32-uart";

etienne-lms commented on boards/st/nucleo_u385rg_q/doc/index.rst:
```suggestion
Flashing an application to Nucleo U385RG
----------------------------------------

etienne-lms commented on boards/st/nucleo_u385rg_q/doc/index.rst:
```suggestion
Default debugger for this board is OpenOCD. It can be used in the usual way.

etienne-lms commented on include/zephyr/dt-bindings/clock/stm32u3_clock.h:
```diff
@@ -0,0 +1,103 @@
+#define MCO_PRE_DIV_2  1
+#define MCO_PRE_DIV_4  2
+#define MCO_PRE_DIV_8  3
+#define MCO_PRE_DIV_16 4
```
Add also the other supported divisors?
```c
#define MCO_PRE_DIV_32  5
#define MCO_PRE_DIV_64  6
#define MCO_PRE_DIV_128 7
#define MCO_PRE_DIV_8  3
#define MCO_PRE_DIV_16 4

etienne-lms commented on include/zephyr/dt-bindings/clock/stm32u3_clock.h:
```diff
@@ -0,0 +1,103 @@
+#define ADF1_SEL(val)		STM32_DT_CLOCK_SELECT((val), 3, 0, CCIPR2_REG)
+#define SPI3_SEL(val)		STM32_DT_CLOCK_SELECT((val), 1, 3, CCIPR2_REG)
+#define SAI1_SEL(val)		STM32_DT_CLOCK_SELECT((val), 3, 5, CCIPR2_REG)
+#define SPI4_SEL(val)		STM32_DT_CLOCK_SELECT((val), 1, 7, CCIPR2_REG)
+#define I2C4_SEL(val)		STM32_DT_CLOCK_SELECT((val), 1, 9, CCIPR2_REG)
```
```suggestion
```
There is no SPI4 nor I2C4 instances.

etienne-lms commented on include/zephyr/dt-bindings/clock/stm32u3_clock.h:
```diff
@@ -0,0 +1,103 @@
+#define ADCDAC_SEL(val)		STM32_DT_CLOCK_SELECT((val), 3, 16, CCIPR2_REG)
+#define DAC1SH_SEL(val)		STM32_DT_CLOCK_SELECT((val), 1, 19, CCIPR2_REG)
+#define OCTOSPI_SEL(val)	STM32_DT_CLOCK_SELECT((val), 1, 20, CCIPR2_REG)
+#define USART2_SEL(val)		STM32_DT_CLOCK_SELECT((val), 1, 22, CCIPR2_REG)
```
```suggestion
```
There is no USART2 instance.

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,572 @@
+#include <stm32_ll_utils.h>
+#include <stm32_ll_system.h>
+#include <zephyr/arch/cpu.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
```
nitpicking suggestion:
```patch
 #include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
 #include <zephyr/sys/util.h>
-#include <zephyr/drivers/clock_control/stm32_clock_control.h>

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,572 @@
+		*rate = get_msis_frequency();
+		break;
+#endif /* STM32_MSIS_ENABLED */
+#if defined(STM32_MSIK_ENABLED)
+	case STM32_SRC_MSIK:
+#ifndef CONFIG_SOC_SERIES_STM32U3X /* TBD */
+		*rate = __LL_RCC_CALC_MSIK_FREQ(LL_RCC_MSIRANGESEL_RUN,
+				STM32_MSIK_RANGE << RCC_ICSCR1_MSIKRANGE_Pos);
+#endif
+		break;
+#endif /* STM32_MSIK_ENABLED */
```
I still think this is fragile. As is, *rate may not be filled and the function would returns with a success status.
I would rather suggest:
```patch
 #if defined(STM32_MSIK_ENABLED)
+#ifndef CONFIG_SOC_SERIES_STM32U3X /* TBD */
	case STM32_SRC_MSIK:
-#ifndef CONFIG_SOC_SERIES_STM32U3X /* TBD */
 		*rate = __LL_RCC_CALC_MSIK_FREQ(LL_RCC_MSIRANGESEL_RUN,
 				STM32_MSIK_RANGE << RCC_ICSCR1_MSIKRANGE_Pos);
-#endif
 		break;
+#endif
 #endif /* STM32_MSIK_ENABLED */
```
or like.

kbidani replied:
the development part has been completed.

etienne-lms commented on boards/st/nucleo_u385rg_q/doc/index.rst:
From the DTS file, UART_3_TX is on PC10.
```c
&usart3 {
	pinctrl-0 = <&usart3_tx_pc10 &usart3_rx_pc11>;

etienne-lms commented on dts/bindings/clock/st,stm32u3-msi-clock.yaml:
```diff
@@ -0,0 +1,35 @@
+  msi-range:
+    default: 0
+    type: int
+    description: MSI clock ranges
```
Need to escape first end-of-line.
Also, I think it's worth to detail a bit the diff between from the 2 12MHz and 24MHz possible selections.
```
    description: |
        MSI clock ranges:
        range 0 around 96 MHz
        range 1 around 48 MHz
        range 2 around 24 MHz from 96MHz oscillator
        range 3 around 12 MHz from 96MHz oscillator
        range 4 around 24 MHz from 24MHz oscillator
        range 5 around 12 MHz from 24MHz oscillator
        range 6 around 6 MHz
        range 7 around 3 MHz

kbidani commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,684 @@
+			LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE2);
+		}
+		/* Enable Epod Booster */
+		LL_RCC_SetEPODBoosterClkSource(LL_RCC_EPODBOOSTCLKSRCE_MSIS);
```
in stm32u3xx_ll_rcc.h there are only:
/** @defgroup RCC_LL_EC_EPODBOOST_CLKSOURCE  EPOD clock source selection
  * @{
  */
#define LL_RCC_EPODBOOSTCLKSRCE_BYPASS          0U
#define LL_RCC_EPODBOOSTCLKSRCE_MSIS            RCC_CFGR4_BOOSTSEL_0
#define LL_RCC_EPODBOOSTCLKSRCE_HSI16           RCC_CFGR4_BOOSTSEL_1
#define LL_RCC_EPODBOOSTCLKSRCE_HSE             ( RCC_CFGR4_BOOSTSEL_1 | RCC_CFGR4_BOOSTSEL_0)

kbidani commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,684 @@
+		LL_RCC_MSIK_Enable();
+		while (LL_RCC_MSIK_IsReady() == 0U) {
+		}
+		LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_MSIS);
```
in stm32u3xx_ll_rcc.h there are only:
/** @defgroup RCC_LL_EC_SYS_CLKSOURCE  System clock switch
  * @{
  */
#define LL_RCC_SYS_CLKSOURCE_MSIS          0U                                  /*!< MSIS selection as system clock    */
#define LL_RCC_SYS_CLKSOURCE_HSI16         RCC_CFGR1_SW_0                      /*!< HSI16 selection as system clock   */
#define LL_RCC_SYS_CLKSOURCE_HSE           RCC_CFGR1_SW_1                      /*!< HSE selection as system clock     */

erwango replied:
ok, Txs.

erwango commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,684 @@
+		LL_RCC_MSIS_Enable();
+		while (LL_RCC_MSIS_IsReady() == 0U) {
+		}
+		LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_MSIS);
+
```
Club this line with related while block
```suggestion

		LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_MSIS);
```

erwango commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,684 @@
+		LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_MSIS);
+
+		/* Wait till System clock is ready */
+		while (LL_RCC_GetSysClkSource() != LL_RCC_SYS_CLKSOURCE_MSIS) {
+		}
+
+		LL_RCC_SetAHBPrescaler(LL_RCC_HCLK_SYSCLK_DIV_1);
+		LL_RCC_SetAPB1Prescaler(LL_RCC_APB1_HCLK_DIV_1);
+		LL_RCC_SetAPB2Prescaler(LL_RCC_APB2_HCLK_DIV_1);
+		LL_RCC_SetAPB3Prescaler(LL_RCC_APB3_HCLK_DIV_1);
+
+		__ASSERT(STM32_LSE_ENABLED,
+			"MSIK Hardware auto calibration needs LSE clock activation");
+		if (IN_RANGE(STM32_MSIK_RANGE, 0, 3)) {
+			/*
+			 * LSE or HSE must be enabled and ready before selecting
+			 * this oscillator as MSIRC0 input clock for Range [0 3]
+			 */
+			if (LL_RCC_LSE_IsEnabled() != 0U && LL_RCC_LSE_IsReady() != 0U) {
+				LL_RCC_MSI_RC0_SetPLLInputClk(LL_RCC_MSIPLL0SEL_LSE);
+			} else if (LL_RCC_HSE_IsEnabled() != 0U && LL_RCC_HSE_IsReady() != 0U) {
+				LL_RCC_MSI_RC0_SetPLLInputClk(LL_RCC_MSIPLL0SEL_HSE_OR_HSEDIV2);
+			}
+			LL_RCC_MSI_RC0_PLLmode_Enable();
+			while (LL_RCC_MSI_RC0_PLLmode_IsEnabled() == 0U) {
+			}
+		} else {
+			/*
+			 * LSE or HSE must be enabled and ready before selecting
+			 * this oscillator as MSIRC1 input clock for Range [4 7]
+			 */
+			if (LL_RCC_LSE_IsEnabled() != 0U && LL_RCC_LSE_IsReady() != 0U) {
+				LL_RCC_MSI_RC1_SetPLLInputClk(LL_RCC_MSIPLL1SEL_LSE);
+			} else if (LL_RCC_HSE_IsEnabled() != 0U && LL_RCC_HSE_IsReady() != 0U) {
+				LL_RCC_MSI_RC1_SetPLLInputClk(LL_RCC_MSIPLL1SEL_HSE_OR_HSEDIV2);
+			}
+			LL_RCC_MSI_RC1_PLLmode_Enable();
+			while (LL_RCC_MSI_RC1_PLLmode_IsEnabled() == 0U) {
+			}
+		}
```
This whole code block is duplicated with MSIS and can be replaced by a function which would take range as argument (the only variable part)

erwango commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,684 @@
+		} else {
+			LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE2);
+		}
+		/* Enable Epod Booster */
+		LL_RCC_SetEPODBoosterClkSource(LL_RCC_EPODBOOSTCLKSRCE_MSIS);
+		LL_RCC_SetEPODBoosterClkPrescaler(LL_RCC_EPODBOOSTCLKPRESCAL_1);
+		LL_PWR_EnableEPODBooster();
+		while (LL_PWR_IsActiveFlag_BOOST() == 0) {
+		}
```
Duplicated code block. You could use a function.

erwango reviewed (changes requested):
Some more comments on the clock driver.

Also please change commit title "drivers: flash: Fix log warning".

Otherwise LGTM

etienne-lms reviewed (commented):
Sorry for this late comments. I forgot to look at the flash driver commit.

> Also please change commit title "drivers: flash: Fix log warning".

I agree, commit header line *"drivers: flash: Fix log warning"*   should rather be somthing like
*"drivers: flash: add stm32u3 devices"*.

erwango commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,647 @@
+		} else {
+			LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE2);
+		}
+		enable_epod_booster();
```
Break a line before this.

erwango commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,692 @@
+		/* Enable Epod Booster */
+		enableEpodBooster();
+
+		/* Activate MSIK as source */
```
Comment should be removed

erwango commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,647 @@
+		LL_RCC_MSIK_Enable();
+		while (LL_RCC_MSIK_IsReady() == 0U) {
+		}
+		configure_clock_with_calibration(STM32_MSIK_RANGE);
```
Breaka line before this

erwango commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,647 @@
+		LL_RCC_MSIS_Enable();
+		while (LL_RCC_MSIS_IsReady() == 0U) {
+		}
+		configure_clock_with_calibration(STM32_MSIS_RANGE);
```
Break a line before this

erwango commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,647 @@
+		} else {
+			LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE2);
+		}
+		enable_epod_booster();
```
Break a line before this.

erwango commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,647 @@
+		}
+		enable_epod_booster();
+
+		/* Activate MSIS as source */
```
Comment should be removed/updated.

erwango reviewed (changes requested):
Last clean up and we're good to go.

etienne-lms commented on drivers/flash/flash_stm32l5x.c:
```diff
@@ -21,7 +21,8 @@ LOG_MODULE_REGISTER(LOG_DOMAIN);
 
-#if defined(CONFIG_SOC_SERIES_STM32H5X) || defined(CONFIG_SOC_SERIES_STM32U5X)
+#if defined(CONFIG_SOC_SERIES_STM32H5X) || defined(CONFIG_SOC_SERIES_STM32U5X) \
+	    || defined(CONFIG_SOC_SERIES_STM32U3X)
```
i think you can discard this change. STM32U3x do not build this source file.

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,649 @@
+	LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_HSE);
+	while (LL_RCC_GetSysClkSource() != LL_RCC_SYS_CLKSOURCE_STATUS_HSE) {
+	}
+#elif STM32_SYSCLK_SRC_MSIS
```
`#elif defined(STM32_SYSCLK_SRC_MSIS)`

etienne-lms commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,649 @@
+	LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_MSIS);
+	while (LL_RCC_GetSysClkSource() != LL_RCC_SYS_CLKSOURCE_STATUS_MSIS) {
+	}
+#elif STM32_SYSCLK_SRC_HSI
```
`#elif defined(STM32_SYSCLK_SRC_HSI)`

erwango reviewed (dismissed):
Ok for me (as long as CI issues are fixed). Good job!

mathieuchopstm reviewed (commented):
Otherwise LGTM

kbidani commented on drivers/flash/flash_stm32u3x.c:
```diff
@@ -0,0 +1,264 @@
+	static struct flash_pages_layout stm32_flash_layout[1];
+
+	if (stm32_flash_layout[0].pages_count == 0) {
+		if (stm32_flash_has_2_banks(dev)) {
+			stm32_flash_layout[0].pages_count = FLASH_PAGE_NB;
+		} else {
+			stm32_flash_layout[0].pages_count = FLASH_PAGE_NB / 2;
+		}
+		stm32_flash_layout[0].pages_size = FLASH_PAGE_SIZE;
```
you are right! in stm32u3xx_hal_falsh.h we have:
#define FLASH_PAGE_NB (FLASH_BANK_SIZE / FLASH_PAGE_SIZE)
This means that FLASH_PAGE_NB represents the number of pages per bank. In other words, if a device has two banks, the total number of pages for the device will be FLASH_PAGE_NB * 2.

etienne-lms replied:
Nice catch. My mistake. Thanks.

kbidani commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,648 @@
+
+	/* Set flash latency */
+	/* If freq increases, set flash latency before any clock setting */
+	if (old_hclk_freq < CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC) {
+		LL_SetFlashLatency(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
+	}
```
U3 series does not use a dedicated FLASH_PRESCALER. Instead the flash latency is directly configured using the  LATENCY bits in FLASH_ACR

mathieuchopstm commented on drivers/clock_control/clock_stm32_ll_u3.c:
```diff
@@ -0,0 +1,648 @@
+	LL_RCC_SetAPB2Prescaler(LL_RCC_APB2_HCLK_DIV_1);
+	LL_RCC_SetAPB3Prescaler(LL_RCC_APB3_HCLK_DIV_1);
+
+	BUILD_ASSERT(STM32_LSE_ENABLED || !IS_ENABLED(STM32_MSIx_ENABLED),
+		"MSIx requires LSE clock to be enabled for auto-calibration");
```
My suggestion wasn't clear, but you need to replace the `x` with instance name :sweat_smile: (unless I missed something and `STM32_MSIx_ENABLED` is a real macro??)
```suggestion
	BUILD_ASSERT(STM32_LSE_ENABLED || !IS_ENABLED(STM32_MSIK_ENABLED),
		"MSIK requires LSE clock to be enabled for auto-calibration");
	BUILD_ASSERT(STM32_LSE_ENABLED || !IS_ENABLED(STM32_MSIS_ENABLED),
		"MSIS requires LSE clock to be enabled for auto-calibration");
```

\+ build assert can be moved at top of file

kbidani replied:
yes, the two macros are defined in stm32_clock_control.h:

#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msik), st_stm32u5_msi_clock, okay) || \
	DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msik), st_stm32u3_msi_clock, okay)
#define STM32_MSIK_ENABLED	1

#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msis), st_stm32u5_msi_clock, okay) || \
	DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msis), st_stm32u3_msi_clock, okay)
#define STM32_MSIS_ENABLED	1

mathieuchopstm reviewed (commented):
Just this tiny thing to fix then LGTM
